mautic/mautic
# Add constrain to check if plugin can be published.

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [N ]
| New feature/enhancement? (use the a.x branch)      | [Y]
| Deprecations?                          | [N]
| BC breaks? (use the c.x branch)        | [N]
| Automated tests included?              | [Y] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:

When authorizing the plugin, if my user doesn't have particular access, I should receive an error and fail to authorize the plugin. This i new constrain which we need to add in any plugin which will not allow user to publish the plugin if user don't have sufficient access.
Also clear the access and refresh token whenever credentials are changed.

Example usage in a plugin:
```php
class CanPublishValidatorSubscriber implements EventSubscriberInterface
{
    public static function getSubscribedEvents()
    {
        return [
            PluginEvents::PLUGIN_IS_PUBLISHED_STATE_CHANGING => ['onPublishStateChange', 0],
        ];
    }

    public function onPublishStateChange(PluginIsPublishedEvent $event)
    {
        if ('YourIntegrationName' !== $event->getIntegrationName() || 1 !== $event->getValue()) {
            return;
        }

        // custom logic here...

        $event->setCanPublish($enabled);
        $event->setMessage($message);
    }
}
```

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Not testable without adding constrain in plugin.
